### ISSUE TITLE
calling a type with overloads on `__new__` is not handled in a consistent way

### ISSUE THREAD

**Issue by OutSquareCapital** [astral-sh/ty#4580]
### Summary

Haven't found a matching issue when checking for overload + call. Apologies if already existing.

Calling a type `Foo` like `Foo(arg)` work without issues, but when the type is provided as a `Callable` argument to another function or method, it doesn't pick the right overload, resulting in erroneous outputs.

## Example

`Vec[T]` implement `collections::abc::MutableSequence[T]`, and `Seq[T]`,  spoilers ahead,  `collections::abc::Sequence[T]`.

Hence the former is/should be invariant, and the latter covariant.

Note that all the typing lives in stub. the source code comes from rust.

```python
import operator
from typing import assert_type

from pyochain import Option, Seq, Some, Vec


def check_constructor_call_context() -> None:
    """Constructors should be handled the same, regardless of the call context.

    Checking with `Seq` and `Vec` to ensure variance doesn't affect the result.

    Warning:
        Currently only passes with `basedpyright`
    """
    x = assert_type(Seq([1]), Seq[int])

    _vec_standard = assert_type(Vec(x), Vec[int])
    _seq_standard = assert_type(Seq(x), Seq[int])

    _vec_called = assert_type(operator.call(Vec, x), Vec[int])
    _seq_called = assert_type(operator.call(Seq, x), Seq[int])

    _vec_piped = assert_type(x.pipe(Vec), Vec[int])
    _seq_piped = assert_type(x.pipe(Seq), Seq[int])

    _seq_collected = assert_type(x.iter().collect(Seq), Seq[int])
    _vec_collected = assert_type(x.iter().collect(Vec), Vec[int])

    x_opt = assert_type(Some(x), Option[Seq[int]])
    _seq_mapped = assert_type(x_opt.map(Vec).unwrap(), Vec[int])
    _vec_mapped = assert_type(x_opt.map(Seq).unwrap(), Seq[int])
```

Both of them have thee exact same constructor signature:
```python
    @overload
    def __new__(cls, /) -> Self: ...
    @overload
    def __new__(cls, data: Iterable[T], /) -> Self: ...
    @overload
    def __new__(cls, data: T, /, *more: T) -> Self: ...
    def __new__(cls, data: Iterable[T] | T = (), /, *more: T) -> Self: """documentation"""
```

`pipe` uses `ParamSpec`, whilst `collect` doesn't. So it doesn't seem to be the cause.

Basedpyright report nothing, which is expected, while pyrefly seems to have the same issues as ty here.

File: https://github.com/OutSquareCapital/pyochain/blob/master/tests/test_typing/check_mixins.py

[`pipe` method](https://github.com/OutSquareCapital/pyochain/blob/43c237777225a861a9994df01321e3e6b35fa642/pyochain/abc/_mixins.pyi#L8)

[`collect` method](https://github.com/OutSquareCapital/pyochain/blob/43c237777225a861a9994df01321e3e6b35fa642/pyochain/abc/_iterator.pyi#L813)

[`ty.toml` config](https://github.com/OutSquareCapital/pyochain/blob/master/ty.toml)

## Command

at the repo root:

```shell
uvx ty check tests/test_typing/check_mixins.py
```

output:
```shell
(pyochain) PS C:\Users\stett\Documents\python\pyochain> uvx ty check tests/test_typing/check_mixins.py
error[type-assertion-failure]: Argument does not have asserted type `Vec[int]`
  --> tests\test_typing\check_mixins.py:20:19
   |
20 |     _vec_called = assert_type(operator.call(Vec, x), Vec[int])
   |                   ^^^^^^^^^^^^---------------------^^^^^^^^^^^
   |                               |
   |                               Inferred type is `Vec[T@Vec]`
info: `Vec[int]` and `Vec[T@Vec]` are not equivalent types

error[type-assertion-failure]: Argument does not have asserted type `Seq[int]`
  --> tests\test_typing\check_mixins.py:21:19
   |
21 |     _seq_called = assert_type(operator.call(Seq, x), Seq[int])
   |                   ^^^^^^^^^^^^---------------------^^^^^^^^^^^
   |                               |
   |                               Inferred type is `Seq[T@Seq]`
info: `Seq[int]` and `Seq[T@Seq]` are not equivalent types

error[type-assertion-failure]: Argument does not have asserted type `Vec[int]`
  --> tests\test_typing\check_mixins.py:23:18
   |
23 |     _vec_piped = assert_type(x.pipe(Vec), Vec[int])
   |                  ^^^^^^^^^^^^-----------^^^^^^^^^^^
   |                              |
   |                              Inferred type is `Unknown`
info: `Vec[int]` and `Unknown` are not equivalent types

error[type-assertion-failure]: Argument does not have asserted type `Seq[int]`
  --> tests\test_typing\check_mixins.py:24:18
   |
24 |     _seq_piped = assert_type(x.pipe(Seq), Seq[int])
   |                  ^^^^^^^^^^^^-----------^^^^^^^^^^^
   |                              |
   |                              Inferred type is `Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])]`
info: `Seq[int]` and `Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])]` are not equivalent types

error[invalid-argument-type]: Argument to bound method `Pipe.pipe` is incorrect
  --> tests\test_typing\check_mixins.py:24:37
   |
24 |     _seq_piped = assert_type(x.pipe(Seq), Seq[int])
   |                                     ^^^ Expected `(Seq[int] | (Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])] & Top[Pipe]), /) -> Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])]`, found `<class 'Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])]'>`
info: type `<class 'Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])]'>` has inferred callable type `Overload[() -> Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])], (data: Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])], /) -> Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])], (data: Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])]), /, *more: Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])) -> Seq[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int] | (Top[Pipe] & Iterable[Seq[int]])])])]]`
info: Method defined here
  --> pyochain\abc\_mixins.pyi:8:9
   |
 8 |     def pipe[**P, R](
   |         ^^^^
 9 |         self,
10 |         func: Callable[Concatenate[Self, P], R],
   |         --------------------------------------- Parameter declared here

error[type-assertion-failure]: Argument does not have asserted type `Seq[int]`
  --> tests\test_typing\check_mixins.py:26:22
   |
26 |     _seq_collected = assert_type(x.iter().collect(Seq), Seq[int])
   |                      ^^^^^^^^^^^^---------------------^^^^^^^^^^^
   |                                  |
   |                                  Inferred type is `Seq[int | Top[Iterator[int]]]`
info: `Seq[int]` and `Seq[int | Top[Iterator[int]]]` are not equivalent types

error[type-assertion-failure]: Argument does not have asserted type `Vec[int]`
  --> tests\test_typing\check_mixins.py:27:22
   |
27 |     _vec_collected = assert_type(x.iter().collect(Vec), Vec[int])
   |                      ^^^^^^^^^^^^---------------------^^^^^^^^^^^
   |                                  |
   |                                  Inferred type is `Vec[int | Iterator[int]]`
info: `Vec[int]` and `Vec[int | Iterator[int]]` are not equivalent types

error[type-assertion-failure]: Argument does not have asserted type `Vec[int]`
  --> tests\test_typing\check_mixins.py:30:19
   |
30 |     _seq_mapped = assert_type(x_opt.map(Vec).unwrap(), Vec[int])
   |                   ^^^^^^^^^^^^-----------------------^^^^^^^^^^^
   |                               |
   |                               Inferred type is `Vec[int | Seq[int]]`
info: `Vec[int]` and `Vec[int | Seq[int]]` are not equivalent types

error[type-assertion-failure]: Argument does not have asserted type `Seq[int]`
  --> tests\test_typing\check_mixins.py:31:19
   |
31 |     _vec_mapped = assert_type(x_opt.map(Seq).unwrap(), Seq[int])
   |                   ^^^^^^^^^^^^-----------------------^^^^^^^^^^^
   |                               |
   |                               Inferred type is `Seq[int] | Seq[Seq[int]]`
info: `Seq[int]` and `Seq[int] | Seq[Seq[int]]` are not equivalent types

Found 9 diagnostics
(pyochain) PS C:\Users\stett\Documents\python\pyochain>
```

### Version

ty 0.0.83 (9c214798c 2026-09-21)

**Comment by carljm** [comment]
Thanks! This is a symptom of #4539
